config/base/derive/src/proxy.rs

```diff
@@ -125,16 +156,27 @@ pub fn impl_load_from_disk(ast: &StructWithFields) -> TokenStream {
     let proxy_name = &ast.ident;
     let disk_trait = quote! { ::iroha_config_base::proxy::LoadFromDisk };
     let error_ty = quote! { ::iroha_config_base::derive::Error };
+    let disk_err_variant = quote! { ::iroha_config_base::derive::Error::DiskError };
+    let serde_err_variant = quote! { ::iroha_config_base::derive::Error::SerdeError };
+    let none_proxy = gen_none_fields_proxy(ast);
     quote! {
         impl #disk_trait for #proxy_name {
-            type Error = #error_ty;
-            fn from_path<P: AsRef<std::path::Path> + std::fmt::Debug + Clone>(path: P) -> Result<Self, Self::Error> {
-                let mut file = std::fs::File::open(path)?;
+            type ReturnValue = Self;
+            fn from_path<P: AsRef<::std::path::Path> + ::std::fmt::Debug + Clone>(path: P) -> Self::ReturnValue {
+                let mut file = ::std::fs::File::open(path).map_err(#disk_err_variant);
                 // String has better parsing speed, see [issue](https://github.com/serde-rs/json/issues/160#issuecomment-253446892)
                 let mut s = String::new();
-                std::io::Read::read_to_string(&mut file, &mut s)?;
-                let res: Self = serde_json::from_str(&s)?;
-                Ok(res)
+                let res = file
+                    .and_then(|mut f| {
+                        ::std::io::Read::read_to_string(&mut f, &mut s).map_err(#disk_err_variant)
+                    })
+                    .and_then(
+                        |_| -> ::core::result::Result<Self, #error_ty> {
+                            serde_json::from_str(&s).map_err(#serde_err_variant)
+                        },
+                    )
```

Comment:
I think it is possible to avoid referencing outer scope inside closure.

```suggestion
                    .and_then(|mut f| {
                        let mut s = String::new();
                        ::std::io::Read::read_to_string(&mut f, &mut s).map(move |_| s).map_err(#disk_err_variant)
                    })
                    .and_then(
                        |s| -> ::core::result::Result<Self, #error_ty> {
                            serde_json::from_str(&s).map_err(#serde_err_variant)
                        },
                    )
```

Reply:
I guess it does lead to a somewhat smaller cognitive load, yeah. Commiting.